sys/proc.c

```diff
@@ -45,3 +54,170 @@ proc_t *proc_find(pid_t pid) {
   }
   return p;
 }
+
+/* Always call proc_reap with parent process locked. */
+static int proc_reap(proc_t *child, int *status) {
+  /* Child is now a zombie. Gather its data, cleanup & free. */
+  mtx_lock(&child->p_lock);
+
+  /* We should have the only reference to the zombie child now, we're about to
+     free it. I don't think it may ever happen that there would be multiple
+     references to a terminated process, but if it does, we would need to
+     introduce refcounting for processes. */
+  *status = child->p_exitstatus;
+
+  int retval = child->p_pid;
+
+  assert(&child->p_parent);
+  assert(mtx_owned(&child->p_parent->p_lock));
+  TAILQ_REMOVE(&child->p_parent->p_children, child, p_child);
+
+  WITH_MTX_LOCK (&zombie_proc_list_mtx)
+    TAILQ_REMOVE(&zombie_proc_list, child, p_zombie);
+
+  kfree(M_PROC, child);
+
+  return retval;
+}
+
+void proc_exit(int exitstatus) {
+  /* NOTE: This is a significantly simplified implementation! We currently
+     assume there is only one thread in a process. */
+
+  thread_t *td = thread_self();
+  proc_t *p = td->td_proc;
+  assert(p);
+
+  WITH_MTX_LOCK (&p->p_lock) {
+
+    assert(p->p_nthreads == 1);
+    /* If the process had other threads, we'd need to wake the sleeping ones,
+       request all of them except this one to call thread_exit from
+       exc_before_leave (using a TDF_? flag), join all of them to wait until
+       they terminate. */
+
+    /* Process orphans. They should become children of PID1, but since we don't
+       use an init process yet, I have to make them parent-less (so that can no
+       longer refer to this terminated process). */
+    proc_t *child;
+    TAILQ_FOREACH (child, &p->p_children, p_child) {
+      WITH_MTX_LOCK (&child->p_lock) {
+        child->p_parent = NULL;
+        /* XXX: Assign to init, and send it a SIGCHLD. */
+      }
+    }
+
+    /* Clean up process resources. */
+    vm_map_delete(p->p_uspace);
+    fdtab_release(p->p_fdtable);
+
+    /* Record some process statistics that will stay maintained in zombie
+       state. */
+    p->p_exitstatus = exitstatus;
+
+    /* Turn the process into a zombie. */
+    WITH_MTX_LOCK (&all_proc_list_mtx)
+      TAILQ_REMOVE(&all_proc_list, p, p_all);
+
+    p->p_state = PRS_ZOMBIE;
+
+    WITH_MTX_LOCK (&zombie_proc_list_mtx)
+      TAILQ_INSERT_TAIL(&zombie_proc_list, p, p_zombie);
+
+    /* Notify the parent, in various ways, about state change. */
+    if (p->p_parent) {
+      WITH_MTX_LOCK (&p->p_parent->p_lock) {
+        sleepq_broadcast(&p->p_parent->p_children);
+        sleepq_broadcast(&p->p_state);
+
+        /* If the parent explicitly ignores SIGCHLD, reap child immediately. */
+        if (p->p_sigactions[SIGCHLD].sa_handler == SIG_IGN) {
+          int ignore_status;
+          proc_reap(p, &ignore_status);
+          thread_exit();
```

Comment:
I'm afraid `noreturn` function called from within `WITH_STMT` will prevent from the lock being released :(